go-controller/pkg/clustermanager/zone_cluster_controller.go

```diff
@@ -0,0 +1,354 @@
+	cache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	objretry "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+const (
+	// Maximum node IDs that can be generated. Limited to maximum nodes supported by k8s.
+	maxNodeIDs = 5000
+)
+
+// zoneClusterController is the cluster controller for managing all the zone(s) in the cluster.
+type zoneClusterController struct {
+	kube         kube.Interface
+	watchFactory *factory.WatchFactory
+	stopChan     chan struct{}
+	wg           *sync.WaitGroup
+
+	// node events factory handler
+	nodeHandler *factory.Handler
+
+	// retry framework for nodes
+	retryNodes *objretry.RetryFramework
+
+	// ID allocator for the nodes
+	idAllocator *IDAllocator
+}
+
+func newZoneClusterController(ovnClient *util.OVNClusterManagerClientset, wf *factory.WatchFactory) (*zoneClusterController, error) {
+	idAllocator, err := NewIDAllocator("NodeIDs", maxNodeIDs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an IdAllocator for the nodes, err: %v", err)
+	}
+
+	// Reserve the id 0. We don't want to assign this id to any of the nodes.
+	if err := idAllocator.ReserveID("zero", 0); err != nil {
+		return nil, fmt.Errorf("idAllocator failed to reserve id 0")
+	}
+
+	kube := &kube.Kube{
+		KClient: ovnClient.KubeClient,
+	}
+	wg := &sync.WaitGroup{}
+	zcc := &zoneClusterController{
+		kube:         kube,
+		watchFactory: wf,
+		stopChan:     make(chan struct{}),
+		wg:           wg,
+		idAllocator:  idAllocator,
+	}
+
+	zcc.initRetryFramework()
+	return zcc, nil
+}
+
+func (zcc *zoneClusterController) initRetryFramework() {
+	// We are interested in only nodes
+	resourceHandler := &objretry.ResourceHandler{
+		HasUpdateFunc:          true,
+		NeedsUpdateDuringRetry: false,
+		ObjType:                factory.NodeType,
+		EventHandler: &zoneClusterControllerEventHandler{
+			objType:  factory.NodeType,
+			zcc:      zcc,
+			syncFunc: nil,
+		},
+	}
+
+	zcc.retryNodes = objretry.NewRetryFramework(zcc.stopChan, zcc.wg, zcc.watchFactory, resourceHandler)
+}
+
+// Start starts the zone cluster controller to watch the kubernetes nodes
+func (zcc *zoneClusterController) Start(ctx context.Context) error {
+	nodeHandler, err := zcc.retryNodes.WatchResource()
+
+	if err != nil {
+		return fmt.Errorf("unable to watch nodes: %w", err)
+	}
+
+	zcc.nodeHandler = nodeHandler
+	return nil
+}
+
+func (zcc *zoneClusterController) Stop() {
+	close(zcc.stopChan)
+	zcc.wg.Wait()
+
+	if zcc.nodeHandler != nil {
+		zcc.watchFactory.RemoveNodeHandler(zcc.nodeHandler)
+	}
+}
+
+// handleAddUpdateNodeEvent handles the add or update node event
+func (zcc *zoneClusterController) handleAddUpdateNodeEvent(node *corev1.Node) error {
+	allocatedNodeID, err := zcc.idAllocator.AllocateID(node.Name)
```

Comment:
If you see this is kind of handled in the function zoneClusterControllerEventHandler.AreResourcesEqual()

First time when a node is added, this function will be called.  So ID needs to be allocated and updated.

But for the node update events, handleAddUpdateNodeEvent() will be called only when the node id is out of sync (for some reason).

I'll see if the UpdateNode can avoided or not.   I still need to address @tssurya  comment on using SetNodeAnnotation() which would use Patch.  So maybe it won't be that expensive.

Reply:
yup I also tested numan's patch on that accord to ensure when kubelet updates the node every 5 mins we aren't doing anything stupid - we are not ; because like numan points out he checks only for annotation updates in the `AreResourcesEqual` func; and I have told him to go a step further to explicitly check only if the annotations we care about on the node change we are taking action...

Reply:
Awesome! TY both.